scripts/benchmark/trainer-benchmark.py

```diff
@@ -0,0 +1,435 @@
+import io
+import itertools
+import json
+import math
+import os
+import platform
+import re
+import shlex
+import subprocess
+import sys
+from pathlib import Path
+from statistics import fmean
+
+import pandas as pd
+import torch
+from tqdm import tqdm
+
+import transformers
+
+
+nan = float("nan")
+
+
+class Tee:
+    """
+    A helper class to tee print's output into a file.
+    Usage:
+    sys.stdout = Tee(filename)
+    """
+
+    def __init__(self, filename):
+        self.stdout = sys.stdout
+        self.file = open(filename, "a")
+
+    def __getattr__(self, attr):
+        return getattr(self.stdout, attr)
+
+    def write(self, msg):
+        self.stdout.write(msg)
+        # strip tqdm codes
+        self.file.write(re.sub(r"^.*\r", "", msg, 0, re.M))
+
+
+def get_orig_cmd(max_width=80, full_python_path=False):
+    """
+    Return the original command line string that can be replayed nicely and wrapped for 80 char width
+    Args:
+        - max_width: the width to wrap for. defaults to 80
+        - full_python_path: whether to replicate the full path or just the last part (i.e. `python`). default to `False`
+    """
+
+    cmd = []
+
+    # deal with critical env vars
+    env_keys = ["CUDA_VISIBLE_DEVICES"]
+    for key in env_keys:
+        val = os.environ.get(key, None)
+        if val is not None:
+            cmd.append(f"{key}={val}")
+
+    # python executable (not always needed if the script is executable)
+    python = sys.executable if full_python_path else sys.executable.split("/")[-1]
+    cmd.append(python)
+
+    # now the normal args
+    cmd += list(map(shlex.quote, sys.argv))
+
+    # split up into up to MAX_WIDTH lines with shell multi-line escapes
+    lines = []
+    current_line = ""
+    while len(cmd) > 0:
+        current_line += f"{cmd.pop(0)} "
+        if len(cmd) == 0 or len(current_line) + len(cmd[0]) + 1 > max_width - 1:
+            lines.append(current_line)
+            current_line = ""
+    return "\\\n".join(lines)
+
+
+def get_base_cmd(args, output_dir):
+
+    # unwrap multi-line input
+    args.base_cmd = re.sub(r"[\\\n]+", " ", args.base_cmd)
+
+    # remove --output_dir if any and set our own
+    args.base_cmd = re.sub("--output_dir\s+[^\s]+", "", args.base_cmd)
+    args.base_cmd += f" --output_dir {output_dir}"
+
+    # ensure we have --overwrite_output_dir
+    args.base_cmd = re.sub("--overwrite_output_dir\s+", "", args.base_cmd)
+    args.base_cmd += " --overwrite_output_dir"
+
+    return [sys.executable] + shlex.split(args.base_cmd)
+
+
+def process_run_single(id, cmd, variation, output_dir, target_metric_key, metric_keys, verbose):
+    # enable to debug everything but the run itself, to do it fast and see the progress
+    if 0:
+        import random
+        from time import sleep
+
+        sleep(0)
+        return dict(
+            {k: random.uniform(0, 100) for k in metric_keys},
+            **{target_metric_key: random.choice([nan, 10.31, 100.2, 55.6666, 222.22222222])},
```

Comment:
I don't really understand where those numbers are coming from. Are they random?

Reply:
This is for debugging the output formatting - i.e. still perfecting the look-n-feel of the result reports, so if you want to tweak it these return immediately and with a few re-runs it'll give different outputs to validate that different numbers will be formatted well.

Reply:
I added a note explain what is it for